File: OS/fcfs2.c
#include <stdio.h>

struct PCB {
    int pid, at, bt, st, ct, wt, tat;
};

int main() {
    printf("Enter count of processes: ");
    int count,totalTAT=0,totalWT=0,avgTAT=0,avgWT=0;
    scanf("%d", &count);

    struct PCB p[count];
    struct PCB temp;

    // Objects initialization
    for (int i = 0; i < count; i++) {
        p[i].pid = i + 1;
    }

    printf("Enter Arrival times of processes:\n");
    for (int i = 0; i < count; i++) {
        scanf("%d", &p[i].at);
    }

    printf("Enter Burst times of processes:\n");
    for (int i = 0; i < count; i++) {
        scanf("%d", &p[i].bt);
    }

    // Sorting according to arrival times using bubble sort
    for (int i = 0; i < count - 1; i++) {
        for (int j = 0; j < count - i - 1; j++) {
            if (p[j].at > p[j + 1].at) {
                temp = p[j];
                p[j] = p[j + 1];
                p[j + 1] = temp;
            }
        }
    }

    int prevTime = 0;
    for (int i = 0; i < count; i++) {
        if (p[i].at > prevTime) {
            p[i].st = p[i].at;
        } else {
            p[i].st = prevTime;
        }

        p[i].ct = p[i].st + p[i].bt;
        p[i].tat = p[i].ct - p[i].at;
        p[i].wt = p[i].tat - p[i].bt;

        prevTime = p[i].ct;
    }

    printf("Processes\tArrivalTime\tBurstTimes\tStartingTime\tCompletionTime\tWaitingTime\tTurnAroundTime\n");
    for (int i = 0; i < count; i++) {
        printf("P%d\t\t%d\t\t%d\t\t%d\t\t%d\t\t%d\t\t%d\n",
               p[i].pid, p[i].at, p[i].bt, p[i].st, p[i].ct, p[i].wt, p[i].tat);
    }
    for(int i=0;i<count;i++)
    {
        totalTAT+=p[i].tat;
        totalWT+=p[i].wt;
    }
    avgTAT=totalTAT/count;
    avgWT=totalWT/count;
    printf("\nAverage Turn Around Time =%d",avgTAT);
    printf("\nAverage Waiting Time =%d",avgWT);

    return 0;
}

File: OS/FCFS2.java
import java.util.Scanner;
class PCB
{
    int pid,at,bt,st,ct,wt,tat;
}
public class FCFS2
{
    public static void main(String args[])
    {
        Scanner s=new Scanner(System.in);
        System.out.println("ENter count of processe:");
        int count=s.nextInt();
        PCB[] p=new PCB[count];
        PCB temp=new PCB();

        //objects initialization
        for(int i=0;i<count;i++)
        {
            p[i]=new PCB();
        }

        System.out.println("ENter Arrival times of processes:");
        for(int i=0;i<count;i++)
        {
            p[i].at=s.nextInt();
            p[i].pid=i+1;
        }

        System.out.println("ENter Burst times of processes:");
        for(int i=0;i<count;i++)
        {
            p[i].bt=s.nextInt();
        }

        //sorting according to arrival times

        for(int i=0;i<count-1;i++)
        {
            for(int j=0;j<count-i-1;j++)
            {
                if(p[j].at>p[j+1].at)
                {
                    temp=p[j];
                    p[j]=p[j+1];
                    p[j+1]=temp;
                }
            }
        }
        int prevTime=0;
        for(int i=0;i<count;i++)
        {
            if(p[i].at>prevTime)
            {
                p[i].st=p[i].at;
            }
            else
            {
                p[i].st=prevTime;
            }
            p[i].ct=p[i].st+p[i].bt;
            p[i].tat=p[i].ct-p[i].at;
            p[i].wt=p[i].tat-p[i].bt;

            prevTime=p[i].ct;
        }

        System.out.println("Processes\tArrivalTime\tBurstTimes\tStratingTime\tCompletionTime\tWaitingtime\tTurnAroundtime");
        for(int i=0;i<count;i++)
        {
            System.out.println("P"+p[i].pid+"\t\t"+p[i].at+"\t\t"+p[i].bt+"\t\t"+p[i].st+"\t\t"+p[i].ct+"\t\t"+p[i].wt+"\t\t"+p[i].tat);
        }

        s.close();
    }
}